pin-project-internal/src/pin_projectable/mod.rs

```diff
@@ -0,0 +1,268 @@
+
+impl Parse for PinProject {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let mut items = vec![];
+        while !input.is_empty() {
+            items.push(input.parse()?);
+        }
+        Ok(PinProject { items })
+    }
+}
+
+fn handle_type(args: TokenStream, item: Item, pinned_drop: Option<ItemFn>) -> Result<TokenStream> {
+    match item {
+        Item::Struct(item) => {
+            ensure_not_packed(&item.attrs)?;
+            Ok(structs::parse(args, item, pinned_drop)?)
+        }
+        Item::Enum(item) => {
+            ensure_not_packed(&item.attrs)?;
+            Ok(enums::parse(args, item, pinned_drop)?)
+        }
+        _ => panic!("Unexpected item: {:?}", item),
+    }
+}
+
+pub(super) fn pin_project(input: TokenStream) -> Result<TokenStream> {
+    let span = span!(input);
+    let items: Vec<Item> = syn::parse2::<PinProject>(input)?.items;
+
+    let mut found_type = None;
+    let mut found_pinned_drop = None;
+
+    for mut item in items {
+        match &mut item {
+            Item::Struct(ItemStruct { attrs, .. }) | Item::Enum(ItemEnum { attrs, .. }) => {
+                if found_type.is_none() {
+                    if let Some(pos) = attrs.iter().position(|a| a.path.is_ident("pin_projectable")) {
+                        // Remove the 'pin_projectable' attribute, to prevent it from
+                        // being parsed again
+                        let attr = attrs.remove(pos);
+                        let args = match attr.parse_meta()? {
+                            Meta::List(l) => l.nested.into_token_stream(),
+                            Meta::Word(_) => TokenStream::new(),
+                            _ => return Err(error!(span!(attr), "invalid arguments"))
+                        };
+
+                        found_type = Some((item.clone(), args));
+                    } else {
+                        return Err(error!(span, "type delcared in pin_project! must have pin_projectable attribute"))
+                    }
+                } else {
+                    return Err(error!(span, "cannot declare multiple types within pinned module"))
+                }
+            },
+            Item::Fn(ref mut fn_) => {
+                if fn_.attrs.find_remove(PINNED_DROP) {
+                    if found_pinned_drop.is_none() {
+                        found_pinned_drop = Some(fn_.clone());
+                    } else {
+                        return Err(error!(span, "cannot declare multiple functions within pinned module"));
+                    }
+                } else {
+                    return Err(error!(span, "only #[pinned_drop] functions cannot be declared within pinend module"));
+                }
+            },
+            _ => return Err(error!(span, "pinned module may only contain a struct/enum with an option #[pinned_drop] function"))
+        }
+    }
+
+    if found_type.is_none() {
+        return Err(error!(span, "pin_project must declare a struct or enum"));
```

Comment:
The next `match` should take care of this check

Reply:
Fixed in #22.